baybe/utils/augmentation.py

```diff
@@ -0,0 +1,242 @@
+"""Utilities related to data augmentation."""
+
+from collections.abc import Sequence
+from itertools import permutations
+
+import pandas as pd
+
+
+def _row_in_df(row: pd.Series | pd.DataFrame, df: pd.DataFrame) -> bool:
+    """Check whether a row is fully contained in a dataframe.
+
+    Args:
+        row: The row to be checked.
+        df: The dataframe to be checked.
+
+    Returns:
+        Boolean result.
+
+    Raises:
+        ValueError: If `row` is a dataframe that contains more than one row.
+    """
+    if isinstance(row, pd.DataFrame):
+        if len(row) != 1:
+            raise ValueError(
+                f"{_row_in_df.__name__} can only be called with pd.Series or "
+                f"pd.DataFrame's that have exactly one row."
+            )
+        row = row.iloc[0]
+
+    row = row.reindex(df.columns)
+    return (df == row).all(axis=1).any()
+
+
+def df_apply_permutation_augmentation(
+    df: pd.DataFrame,
+    columns: Sequence[str],
+    dependents: Sequence[str] | None = None,
```

Comment:
I agree, I've pushed two commits implementing this and more tests, now also for invalid input

I realized that we dont need the distinction between `columns` and `dependents`. Instead now theres one single `columns:Sequence[Sequence[str]]` argument. It will cause the n'th element of each sequence to permute togehter.

E.g. the simple case of two permutation invariant columns `["Sol1", "Sol2"]` would be done like 
```python
df_apply_permutation_augmentation(df, columns=[["Sol1"], ["Sol2"]])
```

In the more complex case where each solvent has dependent process parameters `Temp` and `Press` it would look like:
```python
df_apply_permutation_augmentation(df, columns=[["Sol1", "Temp1", "Press1"], ["Sol2", "Temp2" ,"Press2"]])
```
`Solt1` would be permuted with `Sol2`, `Temp1` with `Temp2` and `Press1` with `Press2`, all in the same permutation order

please check the commits

Reply:
resolving as agreed in meeting